python/cuml/experimental/fil/fil.pyx

```diff
@@ -1228,6 +1263,106 @@ class ForestInference(UniversalBase, CMajorInputTagMixin):
             any power of 2, but little benefit is expected above a chunk size
             of 512.
         """
+        chunk_size = self._get_chunk_size(chunk_size)
         return self.forest.predict(
             X, predict_type="per_tree", preds=preds, chunk_size=chunk_size
         )
+
+    def optimize(
+        self,
+        *,
+        data=None,
+        batch_size=1024,
+        iterations=10,
+        predict_method='predict',
+        max_chunk_size=None,
+        seed=0
+    ):
+        """
+        Find the optimal layout and chunk size for this model
+
+        The optimal value for layout and chunk size depends on the model,
+        batch size, and available hardware. In order to get the most
+        realistic performance distribution, example data can be provided. If
+        it is not, random data will be generated based on the indicated batch
+        size. After finding the optimal layout, the model will be reloaded if
+        necessary. The optimal chunk size will be used to set the default chunk
+        size used if none is passed to the predict call.
+
+        Parameters
+        ----------
+        data
+            Example data of shape iterations x batch size x features or None.
+            If None, random data will be generated instead.
+        batch_size : int
+            If example data is not provided, random data with this many rows
+            per batch will be used.
```

Comment:
Alternative, we could just accept an int as data parameter, which would imply "please generate data for me." This would avoid ambiguous parameterization.

Reply:
I ultimately did not go with this because the same argument applies to the (newly-renamed) `unique_batches` argument. We could say that data can accept an int _or_ a tuple to cover both of those, but then it starts to get cumbersome to determine the actual intent of the caller. Did they e.g. expect that tuple to be converted to an array and passed in as data?

I think the explicitness of keeping these as separate parameters is worthwhile, but it's definitely a balance of considerations. Thoughts?

Reply:
I see. I think the cleanest solution would then be to provide a generator function that can be used as input if the default random_data shapes are not acceptable:
```python
estimator.optimize(data=fil.random_data_generator(batch_size=1024, unique_batches=10))
```
This would also allow the user to provide their own custom generator function that more closely models their data if needed.

Just providing example data could be _equivalent_ to:
```python
estimator.optimize(data=KFold().get_n_splits(example_data))